File: Refresh/CaseyRefresh-master/TableRefresh/Ali/AliRefreshAnimationView.m
//
//  CyRefreshAnimationView.m
//  IOS_B01
//
//  Created by Casey on 22/01/2019.
//  Copyright © 2019 Casey. All rights reserved.
//

#import "AliRefreshAnimationView.h"
#import "UIView+CyRefreshFrame.h"
#import "CyRefreshBaseData.h"
#import "AliRefreshShadowView.h"

@interface AliRefreshAnimationView (){
    
    
    UIImageView *_logImageView;
    UIImageView *_waterImageView;
    
    AliRefreshShadowView *_shadowView;
    
    UIView *_contentView;
    
    CGFloat logStartY;
    CGFloat logEndY;
    
}


@property (nonatomic, assign)CGFloat moveRate;

@end


@implementation AliRefreshAnimationView


- (instancetype)initWithFrame:(CGRect)frame{
    
    
    self = [super initWithFrame:frame];
    
    if (self) {
        
        
        _waterImageView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"refreshWater.png"]];
        _waterImageView.hidden = YES;
        _waterImageView.frame = CGRectMake(0, 0, 10, 10);
        [self addSubview:_waterImageView];
        
        
        _shadowView = [[AliRefreshShadowView alloc] initWithFrame:CGRectMake(0, self.frame.size.height - 10 - 6, 40, 6)];
        _shadowView.backgroundColor = UIColor.clearColor;
        _shadowView.clipsToBounds = NO;
        _shadowView.cy_centerX = self.cy_width/2;
        [self addSubview:_shadowView];
        
        _contentView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, self.frame.size.width, self.frame.size.height - 10)];
        _contentView.clipsToBounds = YES;
        [self addSubview:_contentView];
        
        
        _logImageView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"refreshLog.png"]];//refreshLog.png
        _logImageView.frame = CGRectMake(self.frame.size.width/2 - 29/2, self.frame.size.height, 29, 24);
        [_contentView addSubview:_logImageView];
        

        
        logStartY = _logImageView.frame.origin.y;
        logEndY =  self.frame.size.height/2;
        self.moveRate = (logStartY - logEndY) / self.frame.size.height;
    }
    
    return self;
}



- (void)layoutSubviews {
    
    [super layoutSubviews];
    
}

- (void)animationScrollViewOffsetY:(CGFloat)offsetY {
    

    [self removeShakeAnimation];
    offsetY = 0 - offsetY;
    
    if (offsetY >= self.cy_height) {
        offsetY = self.cy_height;
    }
    
    CGFloat posY = self.moveRate * offsetY;
    _logImageView.cy_y = self.frame.size.height - posY;
    _shadowView.visibleValueY = posY;
    _shadowView.hidden = NO;
}


- (void)animationFlush {
    
    
    [_shadowView animationCirle:0.3];
    
    [UIView animateWithDuration:0.3 animations:^{
        
        self->_logImageView.cy_y = self.frame.size.height/2 - self->_logImageView.cy_height/2 - 6;
        
      
        
    } completion:^(BOOL finished) {
        
        [self shakeAnimation];
        
    }];
    
    
}


- (void)shakeAnimation {
    
    
    //初始化一个动画
    CABasicAnimation *baseAnimation = [CABasicAnimation animation];
    baseAnimation.keyPath = @"transform.rotation.z";
    baseAnimation.duration = 0.25;
    baseAnimation.fromValue = [NSNumber numberWithFloat:M_PI/16];
    baseAnimation.toValue = [NSNumber numberWithFloat:-M_PI/16];
    baseAnimation.timingFunction = [CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionEaseIn];
    baseAnimation.autoreverses = YES;
    baseAnimation.repeatCount = 1000;
    baseAnimation.fillMode = kCAFillModeForwards;
    [_logImageView.layer addAnimation:baseAnimation forKey:@"shakeAnimation"];
    

    
}

- (void)removeShakeAnimation {
    
    [_logImageView.layer removeAnimationForKey:@"shakeAnimation"];
}


- (void)animationEndLoading:(void (^) (void))completion {
    
    
    [self removeShakeAnimation];
    
    [_shadowView animationOval:0.25];
    
    _waterImageView.cy_centerX = self.cy_width/2;
    _waterImageView.cy_y = 10;
    _waterImageView.cy_width = 10;
    _waterImageView.cy_height = 10;
    _waterImageView.hidden = NO;
    
    [UIView animateWithDuration:0.25 animations:^{
        
        self->_logImageView.cy_y = self.frame.size.height;
        self->_waterImageView.frame = CGRectMake(self.cy_width/2-20/2, 5, 20, 20);
        
    } completion:^(BOOL finished) {
        
        if (completion) {
            completion();
        }
        self->_waterImageView.hidden = YES;
    }];
}



@end

File: Refresh/Refresh/Ali/AliRefreshAnimationView.swift
//
//  AliRefreshAnimationView.swift
//  Refresh
//
//  Created by Casey on 12/04/2019.
//  Copyright © 2019 nb. All rights reserved.
//

import UIKit

class AliRefreshAnimationView: UIView {
    
    let _waterImageView = UIImageView() // 水珠
    let _shadowView = AliRefreshShadowView.init(frame: CGRect.init(x: 0, y: 0, width: 40, height: 6))
    
    let _contentView = UIView()
    let _logImageView = UIImageView()
    
    var _moveRate:CGFloat = 0 // 相对偏移比率
    override init(frame: CGRect) {
        super.init(frame: frame)
        
        _waterImageView.image = UIImage.init(named: "refreshWater.png")
        _waterImageView.isHidden = true
        _waterImageView.frame = CGRect.init(x: 0, y: 0, width: 10, height: 10)
        self.addSubview(_waterImageView)
        
        
        _shadowView.clipsToBounds = false
        _shadowView.backgroundColor = .clear
        _shadowView.y = self.height - 10 - 6
        _shadowView.centerX = self.width/2
        self.addSubview(_shadowView)
        
        _contentView.clipsToBounds = true
        _contentView.frame = CGRect.init(x: 0, y: 0, width: self.width, height: self.height - 10)
        self.addSubview(_contentView)
        
        _logImageView.image = UIImage.init(named: "refreshLog.png")
        _logImageView.frame = CGRect.init(x: self.width/2 - 29/2, y: self.height, width: 29, height: 24)
        _contentView.addSubview(_logImageView)
        
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func layoutSubviews() {
        
        super.layoutSubviews()
        if _moveRate == 0 {
            let logStartY = _logImageView.y
            let logEndY = self.height/2
            _moveRate = (logStartY - logEndY)/self.height
        }
    }
    
    // 根据offsetY来设置，子视图的位置
    func animationScrollViewOffsetY(_ offsetY:CGFloat)  {
        
        removeShakeAnimation()
        
        var minOffsetY = 0-offsetY
        if minOffsetY >= self.height {
            minOffsetY = self.height
        }
        
        let posY = _moveRate * minOffsetY
        _logImageView.y = self.height - posY
        _shadowView.visibleValueY = posY
        _shadowView.isHidden = false
        
        
    }
    
    // 准备加载 动画
    func animationStartLoading()  {
        
        _shadowView.animationCirle(0.3)
        
        UIView.animate(withDuration: 0.3, animations: {
            
            self._logImageView.y = self.height/2 - self._logImageView.height/2 - 6
            
        }) { (finish) in
            
            self.shakeAnimation()
        }
        
    }
    
    // 左右摆动动画
    func shakeAnimation()  {
        
        let baseAnimation = CABasicAnimation.init()
        baseAnimation.keyPath = "transform.rotation.z"
        baseAnimation.duration = 0.25
        baseAnimation.fromValue = Double.pi/16
        baseAnimation.toValue = -Double.pi/16
        baseAnimation.timingFunction = CAMediaTimingFunction.init(name: CAMediaTimingFunctionName.easeIn)
        baseAnimation.autoreverses = true
        baseAnimation.repeatCount = 1000
        baseAnimation.fillMode = .forwards
        _logImageView.layer.add(baseAnimation, forKey: "shakeAnimation")
    }
    
    // 移除动画
    func removeShakeAnimation()  {
        
        
        _logImageView.layer.removeAnimation(forKey: "shakeAnimation")
    }
    
    
    // 加载完 结束动画
    let endAnimationTime:CGFloat = 1 // 0.25
    func animationEndLoading(_ completion: (() -> Void)? )  {
        
        removeShakeAnimation()
        
        _shadowView.animationOvalEnd(endAnimationTime)
        
        _waterImageView.centerX = self.width/2
        _waterImageView.y = 10
        _waterImageView.width = 10
        _waterImageView.height = 10
        _waterImageView.isHidden = false
        
        UIView.animate(withDuration: Double(endAnimationTime), animations: {
            
            self._logImageView.y = self.height
            self._waterImageView.frame = CGRect.init(x: self.width/2 - 20/2, y: 5, width: 20, height: 20)
            
        }) { (finish) in
            
            completion?()
            self._waterImageView.isHidden = true
        }
        
    }
}
